src/Chainweb/BlockHeaderDB/PruneForks.hs

```diff
@@ -41,180 +55,199 @@ import System.LogLevel
+--
 pruneForks_
-    :: LogFunctionText
+    :: HasCallStack
+    => LogFunctionText
     -> BlockHeaderDb
-    -> Natural
-    -> (BlockHeader -> Bool -> IO ())
+    -> MaxRank
+    -> MinRank
+    -> (Bool -> BlockHeader -> IO ())
     -> IO Int
-pruneForks_ logg cdb limit callback = do
-
-    -- find all roots at \(maxEntry - limit\)
-    --
-    m <- maxRank cdb
-    let rootHeight = m - min m limit
-
-    !roots <- keys cdb Nothing Nothing
-        (Just $ MinRank $ Min rootHeight)
-        Nothing
-            -- include all block headers in the root set, so that all headers
-            -- are included in the marking phase. PayloadBlockHashes can be
-            -- shared between blocks at different height, so we must make sure
-            -- that they are retained if a root depends on them.
-        streamToHashSet_
-
-    -- Bloom filter for marking block headers
-    --
-    let s = max (int rootHeight + 10 * HS.size roots) (int rootHeight * 2)
-    let (size, hashNum) = BF.suggestSizing s 0.0001
-    !marked <- stToIO $ BF.new (BF.cheapHashes hashNum) size
+pruneForks_ logg _ (MaxRank (Max mar)) _  _
+    | mar <= 1 = 0 <$ logg Warn ("Skipping database prunning for max bound of " <> sshow mar)
+pruneForks_ logg cdb mar mir callback = do
+    logg Debug $ "Prunning block header database for chain " <> sshow (_chainId cdb)
+        <> " with upper bound " <> sshow (_getMaxRank mar)
+        <> " and lower bound " <> sshow (_getMinRank mir)
+
+    !pivots <- entries cdb Nothing Nothing (Just $ MinRank $ Min $ _getMaxRank mar) (Just mar)
+        $ fmap (force . L.nub) . S.toList_ . S.map _blockParent
+            -- the set of pivots is expected to be very small. In fact it is
+            -- almost always a singleton set.
+
+    if null pivots
+      then do
+        logg Warn
+            $ "Skipping database pruning because of an empty set of block headers at upper pruning bound " <> sshow mar
+        return 0
+      else
+        withReverseHeaderStream cdb (mar - 1) mir
+            $ S.foldM_ go (return (pivots, int (_getMaxRank mar), 0)) (\(_,_,!n) -> evaluate n)
 
-    -- Bloom filter for marking block payload hashes
-    --
-    -- Payloads can be shared between different blocks. Thus, to be on the safe
-    -- side, we are not deleting any marked payload.
-    --
-    -- Note that we could use cuckoo hashes or counting bloom filters to do
-    -- reference counting and collect more unreferenced payloads.
-    --
-    -- TODO: would it be better to use a single shared filter?
-    --
-    let (psize, phashNum) = BF.suggestSizing s 0.0001
-    !markedPayloads <- stToIO $ BF.new (BF.cheapHashes phashNum) psize
+  where
+    go :: ([BlockHash], BlockHeight, Int) -> BlockHeader -> IO ([BlockHash], BlockHeight, Int)
+    go ([], _, _) cur = throwM $ InternalInvariantViolation
+        $ "PrunForks.pruneForks_: no pivots left at block " <> encodeToText (ObjectEncoded cur)
+    go (!pivots, !prevHeight, !n) !cur
```

Comment:
What is the purpose of `n`? It is not read in `go`

Reply:
It counts the number of deleted blocks, which is reported in the logs.

Reply:
`n` is returned by the fold in line 185: `(\(_,_,!n) -> evaluate n)`